### PR TITLE
Vat perturbation

### PR DESCRIPTION
**Pull request by aam-at**
Adds perturbation used for virtual adversarial training as described in https://arxiv.org/abs/1507.00677. Both theano and tensorflow are supported.

### CODE REVIEW

**Comment by aam-at** [comment]
Tests fail for tensorflow on `Failure: ImportError (No module named 'PyQt4')`.

**Comment by npapernot** [comment]
Thanks for the pr!

This test error will be solved after #75 is merged (the pr #75 still needs
a few edits and it will be good to go).

Traveling---will get back to you in a few days

**Comment by npapernot** [comment]
hi @aam-at , would you mind resolving the conflicts with the master branch now that #75 was merged, your PR should pass the tests

**Comment by aam-at** [comment]
@npapernot I resolved conflicts and tests are passing.

**Comment by npapernot** [comment]
Great, thank you! 

Could you take a look @takerum?

**Comment by takerum** [comment]
LGTM!

**Comment by npapernot** [comment]
Thanks @takerum !

File: cleverhans/attacks_tf.py [hunk 1]
```diff
@@ -82,6 +82,39 @@ def fgm(x, preds, y=None, eps=0.3, ord=np.inf, clip_min=None, clip_max=None):
     return adv_x
 
 
+def vatm(model, x, predictions, eps, num_iterations=1, xi=1e-6,
+         clip_min=None, clip_max=None, scope=None):
+    """
+    Tensorflow implementation of the perturbation method used for virtual
+    adversarial training: https://arxiv.org/abs/1507.00677
+    :param model: the model which returns the network unnormalized logits
```
Comment (npapernot) [inline cleverhans/attacks_tf.py]: Same than above (logits instead of model?)

File: cleverhans/utils_tf.py [hunk 2]
```diff
@@ -300,3 +302,42 @@ def model_argmax(sess, x, predictions, samples):
...
+    """
+    with tf.name_scope(name, "l2_normalize", [x]) as name:
+        x = tf.convert_to_tensor(x, name="x")
+        x /= (epsilon + tf.reduce_max(tf.abs(x), dim, keep_dims=True))
+        square_sum = tf.reduce_sum(tf.square(x), dim, keep_dims=True)
+        x_inv_norm = tf.rsqrt(np.sqrt(epsilon) + square_sum)
+        return tf.multiply(x, x_inv_norm, name)
+
+
+def normalize_perturbation(x, name=None):
```
Comment (npapernot) [inline cleverhans/utils_tf.py]: Is it necessary to have two functions `l2_normalize` and `normalize_perturbation`? It looks like they could be merged.

If we keep the two, could we find a more descriptive name than `normalize_perturbation` to indicate that it uses the l2 norm?

File: cleverhans/utils_th.py [hunk 3]
```diff
@@ -285,3 +284,36 @@ def model_argmax(x, predictions, sample):
...
+    :return: the l2 normalized vector
+    """
+    epsilon = np.asarray(epsilon, dtype=floatX)
+    x /= (epsilon + T.max(T.abs_(x), axis, keepdims=True))
+    square_sum = T.sum(T.sqr(x), axis, keepdims=True)
+    x /= T.sqrt(np.sqrt(epsilon) + square_sum)
+    return x
+
+
+def normalize_perturbation(x):
```
Comment (npapernot) [inline cleverhans/utils_th.py]: Same comments than for the TF version

**Comment by npapernot** [comment]
@aam-at Did you get a chance to look at this again?

**Comment by aam-at** [comment]
@npapernot Thanks! I reviewed your comments but didn't have time to make changes. I am traveling right now and I will address your comments as soon as possible.

**Comment by npapernot** [comment]
Just wanted to check in, thanks a lot for your contributions @aam-at!

File: cleverhans/attacks.py [hunk 4]
```diff
@@ -446,6 +446,40 @@ def fgsm(x, predictions, eps, back='tf', clip_min=None, clip_max=None):
         return fgm(x, predictions, eps, clip_min=clip_min, clip_max=clip_max)
 
 
+def vatm(model, x, predictions, eps, back='tf', num_iterations=1, xi=1e-6,
```
Comment (aam-at) [inline cleverhans/attacks.py]: Fixed.

File: cleverhans/attacks.py [hunk 5]
```diff
@@ -446,6 +446,40 @@ def fgsm(x, predictions, eps, back='tf', clip_min=None, clip_max=None):
...
+         clip_min=None, clip_max=None):
+    """
+    A wrapper for the perturbation methods used for virtual adversarial
+    training : https://arxiv.org/abs/1507.00677
+    It calls the right function, depending on the
+    user's backend.
+    :param model: the model which returns the network unnormalized logits
```
Comment (aam-at) [inline cleverhans/attacks.py]: `model` should be a keras model or function which returns the output tensor. I renamed though `predictions` to `logits` as it is necessary for vat to have unnormalized output tensor.

**Comment by aam-at** [comment]
@npapernot Thank you for your review! I addressed your comments.

**Comment by npapernot** [comment]
Thank you for making the changes required, the code matches the API better now

**Comment by aam-at** [comment]
@npapernot Thanks for the updates!

**Comment by npapernot** [comment]
One last thing before we merge @aam-at, we are striving to add more tests to the core module components. Since this will be inside `cleverhans/`, could you write tests in `tests_tf/` that we can rely on to make sure that future PRs don't break your code?

File: cleverhans/attacks_tf.py [hunk 6]
```diff
@@ -82,6 +82,39 @@ def fgm(x, preds, y=None, eps=0.3, ord=np.inf, clip_min=None, clip_max=None):
     return adv_x
 
 
+def vatm(model, x, logits, eps, num_iterations=1, xi=1e-6,
```
Comment (npapernot) [inline cleverhans/attacks_tf.py]: Can you add tests for this function?

File: cleverhans/utils_tf.py [hunk 7]
```diff
@@ -330,3 +332,34 @@ def model_argmax(sess, x, predictions, samples):
         return np.argmax(probabilities)
     else:
         return np.argmax(probabilities, axis=1)
+
+
+def l2_batch_normalize(x, epsilon=1e-12, scope=None):
```
Comment (npapernot) [inline cleverhans/utils_tf.py]: Can you add tests for this function?

File: cleverhans/utils_tf.py [hunk 8]
```diff
@@ -330,3 +332,34 @@ def model_argmax(sess, x, predictions, samples):
...
+        x_shape = tf.shape(x)
+        x = slim.flatten(x)
+        x /= (epsilon + tf.reduce_max(tf.abs(x), 1, keep_dims=True))
+        square_sum = tf.reduce_sum(tf.square(x), 1, keep_dims=True)
+        x_inv_norm = tf.rsqrt(np.sqrt(epsilon) + square_sum)
+        x_norm = tf.multiply(x, x_inv_norm)
+        return tf.reshape(x_norm, x_shape, scope)
+
+
+def kl_with_logits(q_logits, p_logits, scope=None,
```
Comment (npapernot) [inline cleverhans/utils_tf.py]: Can you add tests for this function?

File: cleverhans/attacks.py [hunk 9]
```diff
@@ -412,6 +412,90 @@ def parse_params(self, theta=1., gamma=np.inf, nb_classes=10, clip_min=0.,
         return True
 
 
+class VirtualAdversarialMethod(Attack):
```
Comment (npapernot) [inline cleverhans/attacks.py]: Can you add at least a test that makes sure the class can be instantiated without crashing?

Comment (aam-at) [inline cleverhans/attacks.py]: Fixed.

File: cleverhans/utils_tf.py (hunk 8 above)
Comment (aam-at) [inline cleverhans/utils_tf.py]: Fixed.

File: cleverhans/utils_tf.py (hunk 7 above)
Comment (aam-at) [inline cleverhans/utils_tf.py]: Fixed.

File: cleverhans/attacks_tf.py (hunk 6 above)
Comment (aam-at) [inline cleverhans/attacks_tf.py]: Fixed.

**Comment by aam-at** [comment]
@npapernot I added basic tests for `utils` and `vatm` method. Do you have a roadmap for this project? Can you point me out to the discussion why you are dropping `theano` support? Thanks.

**Comment by goodfeli** [comment]
The discussion was on the cleverhans-dev mailing list "dropping Theano
 support".

On Thu, Jun 1, 2017 at 6:31 AM, Alexander Matyasko <notifications@github.com

**Comment by aam-at** [comment]
@npapernot Thanks for the merge!
@goodfeli Thanks for pointing out to the discussion!

**Comment by npapernot** [comment]
Thanks for the contribution @aam-at, very much appreciated! Can you let us know what your affiliation is so we can add you to the  technical report when we increment the version number of the library?

**Comment by aam-at** [comment]
@npapernot I contacted you via email.
